cmd/forwarder/auth.go

```diff
@@ -0,0 +1,225 @@
+package main
+
+import (
+	"crypto/hmac"
+	"crypto/sha512"
+	"encoding/hex"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"net/http"
+	"net/url"
+	"reflect"
+	"strconv"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/go-redis/redis"
+	metrics "github.com/rcrowley/go-metrics"
+	log "github.com/sirupsen/logrus"
+)
+
+func newAuth(config AuthConfig, registry metrics.Registry) (*BasicAuth, error) {
```

Comment:
Generally commentary on this function ... You may want to have two functions, one for handling auth based on config.Tokens and another for auth based on redis. And then wrap them in a single function.

Reply:
@freeformz I'm working on this one.

Reply:
@freeformz so, I have some thoughts here. 

I originally started out attempting to have multiple objects, each with their own `Authenticate` method. I ran into a few problems doing it. Basically:

If I have multiple invocations of an `Authenticate(r *http.Request)` function (eg: one for creds that come from the environment and another for creds from redis), then under-the-hood, each will need to invoke `hmacEncode` on the input password. This is expensive.

There are ways to mitigate this expense: 

1. I could cache the result of hmacEncode in thread local storage. But if I do this, I must be careful: A brute-force password attack could cause us to cache the results of invalid passwords forever. I need to ensure the cache has an upper size bound, and evicts negative results eventually. I'm not sure what to use to for that. I'm also not sure how to do thread-local storage in go at this point.

2. I could implement a different interface like `Authenticate(user string, passwordHash string), and just have an object that contains a map of strings to a list of `credential` structs.  However, with that implementation I'm starting to feel like there's very little logic in the methods specific to this struct, and it's really just a wrapper around the map - in which case I feel like it's over-engineered and not adding alot of value to the code.

Do you have any suggestions for how I should proceed? If I don't do this now, I think it's unlikely that I will ever do it.

Reply:
@freeformz my preference for right now, sadly, is to not do it, unless you feel strongly. If I don't deploy this during my day tomorrow, I'm unlikely to be able to deploy it until next week.

Reply:
Sorry, I may not have been concrete enough in my comment. I was trying to say:
```go
func newAuth(...) (*BasicAuth, error) {
// config checks
....

result, err := NewBasicAuthFromString(config.Tokens, config.HmacKey, registry)
if err != nil {
  return result, err
}

if config.RedisUrl != "" {
  go refreshAuthFromRedis(result, registry, config.RedisUrl, config.HmacKey, config.RedisKey, config.Tokens)
}
return result, err
```

Basically fully split it up into two functions, each focused on their own source. The one that deals with config.Tokens source already exists in the form of NewBasicAuthFromString.